chapter/background.md

```diff
@@ -0,0 +1,607 @@
+In return, clients have to follow the calling conventions for the requested method.
+Typically, messages are synchronous, i.e., the client awaits a result associated to the method and its parameters before it continues its execution of the calling environment.
+Hence, requests are marked with an `id` which is included in the result/error message necessarily sent by the server.
+If the client does not require a response, it can omit the `id` field.
+The message is then interpreted as a *notification*, which servers cannot respond to.
+
+Part of the JSON-RPC specification is the ability for clients to batch requests and send a list of request or notification objects.
+In this case, the server should respond with a list of results matching each request, yet is free to process requests concurrently.
+
+### Commands and Notifications
+
+The LSP builds on top of the JSON-RPC protocol described in the previous subsection.
+In total the LSP defines 33 [@lsp] "language features", i.e., source code related capabilities.
+In addition, the LSP specifies different capabilities to the server to control the editor.
+For instance, servers may instruct clients to show notifications or progress bars or open documents.
+Similarly, the client has multiple ways of notifying the server of file changes, including renaming r deletion of files.
+
+This thesis aims to implement a fundamental set of capabilities.
+The chosen capabilities are based on those identified as "key methods" by the authors of langserver [@langserver], specifically:
+
+
+1. Code completion
+   Suggest identifiers, methods or values at the cursor position.
+2. Hover information
+   Present additional information about an item under the cursor, i.e., types, contracts and documentation.
+3. Jump to definition
+   Find and jump to the definition of a local variable or identifier.
+4. Find references
+   List all usages of a defined variable.
+5. Workspace/Document symbols
+   List all variables in a workspace or document.
+6. Diagnostics
+   Analyze source code, i.e., parse and type check and notify the LSP Client if errors arise.
+
+#### Code Completion
+
+This feature allows users to request a suggested identifier of variables or methods, concrete values or larger templates of generated code to be inserted at the position of the cursor.
+The completion can be invoked manually or upon entering language defined trigger characters, such as `.`, `::` or `->`.
+The completion request contains the current cursor position, allowing the language server to resolve contextual information based on an internal representation of the document.
+In the example ([@fig:lsp-capability-complete]) the completion feature suggests related identifiers for the incomplete function call "`pr`*`int`*".
+
+![](examples/complete.png){#fig:lsp-capability-hover caption="Completion options resolved by the Python language server in Visual Studio Code"}
```

Comment:
fundamental limitation of reviewing a PR instead of reviewing a PDF, I don't see the figure inline.